### PR TITLE
OCPBUGS-24705: consider currentImage and desiredImage annotations

### PR DESCRIPTION
**Pull request by cheesesashimi**
**- What I did**

The `isNodeUnavailable()` code was not taking the `currentImage` / `desiredImage` annotations into consideration when it determines if the node should be considered unavailable. While opting into on-cluster builds, the following would happen:

1. A node gets updated with the `desiredImage` annotation, indicating that the MCD should transition it to an image.
2. There is a lag between when this annotation is applied and the MCD transitions from "Done" -> "Working".
3. During this lag period, the NodeController may re-enter the sync loop.
4. Because the MCD state has not transitioned yet to "Working" yet and we were not considering the currentImage / desiredImage annotations, the NodeController will apply the `desiredImage` annotation to another node instead.

This adds a check for the presence of the `currentImage` / `desiredImage` annotations. If neither annotation is present, the NodeController will not take these annotations into consideration. However, if either of them is present, it will test their equality before making a determination that a node is "Done". This PR also adds additional test cases in an attempt to verify additional unforeseen edge-cases.

**- How to verify it**

1. Bring up an OpenShift 4.16 cluster.
2. Opt into on-cluster builds using either the docs within this repo or my `onclustertesting` helpers.
3. Opt the `master` pool into on-cluster builds by adding the appropriate label: `$ oc label mcp/master 'machineconfiguration.openshift.io/layering-enabled='`.
4. Only one control plane node should go offline at once.

**- Description for the changelog**
Consider currentImage / desiredImage annotations when determining if a node is unavailable

### CODE REVIEW

File: pkg/controller/common/layered_node_state.go [hunk 1]
```diff
@@ -131,16 +131,66 @@ func isNodeDone(node *corev1.Node) bool {
 	if node.Annotations == nil {
 		return false
 	}
+
+	if !isNodeConfigDone(node) {
```
Comment (cdoern) [inline pkg/controller/common/layered_node_state.go:135]:
```
if !isNodeConfigDone(node) || !isNodeImageDone(node) || !IsNodeMCDState(node, daemonconsts) {
     return false
}
````

**Review by cdoern — APPROVED** [review]
looks good! just one nit

File: pkg/controller/common/layered_node_state.go (hunk 1 above)
Comment (cheesesashimi) [reply pkg/controller/common/layered_node_state.go:135]: I'd rather leave it the way that it is, only because it's pretty explicit about when we return false. Elsewhere in the code, we did something like this and I think that's how this particular bug was introduced.

**Review by yuqi-zhang — APPROVED** [review]
/lgtm

Logic is very clear, and great test coverage!

**Comment by yuqi-zhang** [comment]
/override ci/prow/e2e-gcp-op-single-node

pre-emptively
